plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergSplitSource.java

```diff
@@ -94,7 +116,53 @@ private ConnectorSplit toIcebergSplit(FileScanTask task)
                 task.length(),
                 task.file().fileSizeInBytes(),
                 task.file().format(),
-                ImmutableList.of(),
+                getAddresses(task),
                 getPartitionKeys(task));
     }
+
+    private List<HostAddress> getAddresses(FileScanTask task)
+    {
+        if (fetchSplitLocations) {
+            return blockLocations(task);
+        }
+        else {
+            return ImmutableList.of();
+        }
+    }
+
+    private List<HostAddress> blockLocations(FileScanTask task)
+    {
+        Set<HostAddress> locationSets = new HashSet<>();
+        Path path = new Path(task.file().path().toString());
+        try {
+            FileSystem fs = hdfsEnvironment.getFileSystem(hdfsContext, path);
+            for (BlockLocation blockLocation : fs.getFileBlockLocations(path, task.start(), task.length())) {
+                locationSets.addAll(getHostAddresses(blockLocation));
+            }
+        }
+        catch (IOException e) {
+            log.warn("Failed to get block locations for path %s", path, e);
+        }
+        ArrayList<HostAddress> list = new ArrayList<>(locationSets);
+        return list;
+    }
+
+    private static List<HostAddress> getHostAddresses(BlockLocation blockLocation)
+    {
+        // Hadoop FileSystem returns "localhost" as a default
+        return Arrays.stream(getBlockHosts(blockLocation))
+                .map(HostAddress::fromString)
+                .filter(address -> !address.getHostText().equals("localhost"))
```

Comment:
It's slightly better to put the constant on the left to avoid null pointer exceptions: `!"localhost".equals(addresss.getHostText())`.